src/messages/mod.rs

```diff
@@ -58,6 +60,67 @@ impl HopMessage {
     }
 }
 
+#[derive(Eq, PartialEq, Clone)]
+pub struct HopMessageWithSerializedMessage {
+    /// Wrapped signed message.
+    content: SignedRoutingMessage,
+    /// Serialized Message as received or sent to quic_p2p.
+    serialized_message: Bytes,
+}
+
+impl HopMessageWithSerializedMessage {
+    /// Serialize message and keep both SignedRoutingMessage and Bytes.
+    pub fn new(content: SignedRoutingMessage) -> Result<Self> {
+        let hop_msg = HopMessage::new(content)?;
+        let message = Message::Hop(hop_msg);
+        let serialized_message = to_network_bytes(&message)?;
+
+        let content = if let Message::Hop(hop_msg) = message {
+            hop_msg.content
+        } else {
+            unreachable!("Created as Hop can only match Hop.")
+        };
+
+        Ok(Self {
+            content,
+            serialized_message,
+        })
+    }
+
+    /// Create from part with serialized_message the bytes of the serialization
+    /// for `Message::Hop(HopMessage::new(content))`
+    pub fn new_with_serialized_message(
+        content: SignedRoutingMessage,
+        serialized_message: Bytes,
+    ) -> Self {
+        Self {
+            content,
+            serialized_message,
+        }
+    }
+
+    pub fn signed_routing_message(&self) -> &SignedRoutingMessage {
+        &self.content
+    }
+
+    pub fn into_signed_routing_message(self) -> SignedRoutingMessage {
+        self.content
+    }
+
+    pub fn routing_message(&self) -> &RoutingMessage {
+        self.content.routing_message()
+    }
+
+    pub fn serialized_message(&self) -> Bytes {
+        // Bytes is backed by Rc so clone is cheap and result in the same memory held.
+        self.serialized_message.clone()
```

Comment:
Why not let the caller do the clone if they need to?

Reply:
Because they pretty much always do, and I thought putting the comment here made more sense than having clone where we call it. Not sure I'm right.

Reply:
I feel like letting the caller do the clone is slightly more idiomatic due to being more explicit. Similar to how I think it's better to take argument by value instead of taking it by reference and cloning internally. But I don't feel too strongly about this, so I'll leave it up to you.

Reply:
Yes, the problem is normally with Rc, we would use `Rc::clone()` (clippy suggest it I think), as that highlight it is a different kind of clone, but Bytes does not have it, so I'll probably keep it.

Reply:
Yes, but `Bytes` isn't alias for `Rc`, it just contains one (actually I think it is `Arc`), so calling `clone` on it is perfectly fine even for clippy.

Reply:
Ok, ideally I think it would be best if there was a `bytes::clone()` to make things clear.
Trying to work a way that would address both our concerns.

Option A:
Add a function `bytes_clone` somewhere with explanation that it is cheap and use it at the caller site of serialized_message.

Option B:
Have 2 function `serialized_message` not cloning, and another one doing cloning but also somewhat clear that it is a different clone. I don't really have a good idea.


Would option A makes sense to you, or do you have a good name for option B?

Reply:
What is wrong with just returning `&Bytes` from `serialized_message` and letting the caller call `clone` on it? Is it because it's not obvious that it is a cheap clone?

Reply:
yes

Reply:
maybe that is not as important as I make it to be, but we seem to be trying to avoid unnecessary clones generally.

Reply:
Ok. I'm not sure this is that important. Maybe just add a comment explaining the clone is cheap?

Reply:
Ok, I'll do that at each call site then